nixos/modules/services/web-servers/caddy.nix

```diff
@@ -5,6 +5,24 @@ with lib;
 let
   cfg = config.services.caddy;
   configFile = pkgs.writeText "Caddyfile" cfg.config;
+
+  # v2-specific options
+  isCaddy2 = versionAtLeast cfg.package.version "2.0";
+  tlsConfig = {
+    apps.tls.automation.policies = [{
+      issuer = {
+        inherit (cfg) ca email;
+        module = "acme";
+      };
+    }];
+  };
+  adaptedConfig = importJSON (pkgs.runCommand "caddy-config-adapted.json" { } ''
+    ${cfg.package}/bin/caddy adapt \
+      --config ${configFile} --adapter ${cfg.adapter} > $out
+  '');
+  # TODO: validate with `caddy validate`?
```

Comment:
```suggestion
```

Reply:
to add on, `adapt` validates the input config by default and will return error 1 when encounter invalid input.